### PR TITLE
Add nogil to CUB

### PR DESCRIPTION
**Pull request by y1r**
This PR adds `nogil` flag to CUB C interface.

### CODE REVIEW

File: cupy/cuda/cub.pyx [hunk 1]
```diff
@@ -122,14 +122,28 @@ def device_reduce(ndarray x, op, out=None, bint keepdims=False):
...
-                                                   op, dtype_id)
+    x_size = <int>x.size
+
+    with nogil:
+        ws_size = cub_device_reduce_get_workspace_size(x_ptr,
+                                                       y_ptr,
+                                                       x_size,
+                                                       s,
+                                                       op,
+                                                       dtype_id)
```
Comment (niboshi) [inline cupy/cuda/cub.pyx]: You don't have to use `with nogil:` unless the function call takes long time.
(Ditto for other instances)

File: cupy/cuda/cub.pyx [hunk 2]
```diff
@@ -122,14 +122,28 @@ def device_reduce(ndarray x, op, out=None, bint keepdims=False):
         y = ndarray((kv_bytes,), numpy.int8)
     x_ptr = <void *>x.data.ptr
     y_ptr = <void *>y.data.ptr
-    dtype_id = _get_dtype_id(x.dtype)
+    dtype_id = <int>_get_dtype_id(x.dtype)
```
Comment (niboshi) [inline cupy/cuda/cub.pyx]: This cast is not necessary, as `dtype_id` is already declared `int`.

**Comment by y1r** [comment]
@niboshi I updated my PR. Could you check the PR again?

File: cupy/cuda/cub.pyx [hunk 3]
```diff
@@ -124,12 +124,20 @@ def device_reduce(ndarray x, op, out=None, bint keepdims=False):
     y_ptr = <void *>y.data.ptr
     dtype_id = _get_dtype_id(x.dtype)
     s = <Stream_t>stream.get_current_stream_ptr()
-
+    x_size = <int>x.size
```
Comment (leofang) [inline cupy/cuda/cub.pyx]: As a good Cython coding practice, could you declare `cdef int x_size` on the top?

File: cupy/cuda/cub.pyx [hunk 4]
```diff
@@ -124,12 +124,20 @@ def device_reduce(ndarray x, op, out=None, bint keepdims=False):
...
-    cub_device_reduce(ws_ptr, ws_size, x_ptr, y_ptr, x.size, s, op, dtype_id)
+    with nogil:
+        cub_device_reduce(ws_ptr,
+                          ws_size,
+                          x_ptr,
+                          y_ptr,
+                          x_size,
+                          s,
+                          op,
+                          dtype_id)
```
Comment (leofang) [inline cupy/cuda/cub.pyx]: This line breaking is horrible. Any reason the style can't be preserved? This should be better:
```python
# I didn't do the character count, so this breaking point might be bad
with nogil:
    cub_device_reduce(ws_ptr, ws_size, x_ptr, y_ptr, x_size, s, 
                      op, dtype_id)  
```

File: cupy/cuda/cub.pyx (hunk 3 above)
Comment (niboshi) [reply cupy/cuda/cub.pyx]: Or, if `with nogil:` below can be removed, you can simply revert to inlined `x.size` without casting.
How about removing `with nogil:`?

File: cupy/cuda/cub.pyx [hunk 5]
```diff
@@ -202,9 +210,10 @@ def device_segmented_reduce(ndarray x, op, axis, out=None,
         op, dtype_id)
     ws = memory.alloc(ws_size)
     ws_ptr = <void*>ws.ptr
-    cub_device_segmented_reduce(ws_ptr, ws_size, x_ptr, y_ptr, n_segments,
-                                offset_start_ptr, offset_end_ptr, s,
-                                op, dtype_id)
+    with nogil:
+        cub_device_segmented_reduce(ws_ptr, ws_size, x_ptr, y_ptr, n_segments,
```
Comment (niboshi) [inline cupy/cuda/cub.pyx]: Ditto. How about removing `with nogil:`?

File: cupy/cuda/cub.pyx [hunk 6]
```diff
@@ -258,9 +267,10 @@ def device_csrmv(int n_rows, int n_cols, int nnz, ndarray values,
         n_cols, nnz, s, dtype_id)
     ws = memory.alloc(ws_size)
     ws_ptr = <void *>ws.ptr
-    cub_device_spmv(ws_ptr, ws_size, values_ptr, row_offsets_ptr,
-                    col_indices_ptr, x_ptr, y_ptr, n_rows, n_cols, nnz, s,
-                    dtype_id)
+    with nogil:
+        cub_device_spmv(ws_ptr, ws_size, values_ptr, row_offsets_ptr,
```
Comment (niboshi) [inline cupy/cuda/cub.pyx]: Ditto. How about removing `with nogil:`?

File: cupy/cuda/cub.pyx (hunk 5 above)
Comment (y1r) [reply cupy/cuda/cub.pyx]: `cub_device_segmented_reduce` is actually batched reduce method so I want to add `nogil`.

File: cupy/cuda/cub.pyx (hunk 6 above)
Comment (y1r) [reply cupy/cuda/cub.pyx]: `cub_device_segmented_reduce` is actually a sparse Matrix-Vector multiplication method so I want to add `nogil`.

File: cupy/cuda/cub.pyx (hunk 3 above)
Comment (y1r) [reply cupy/cuda/cub.pyx]: I want to keep `with nogil` so I added `cdef int x_size`

**Comment by niboshi** [comment]
Thank you for fixes.
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

File: cupy/cuda/cub.pyx [hunk 7]
```diff
@@ -91,10 +91,10 @@ cpdef _preprocess_array(ndarray arr, axis, bint keepdims, str order):
     return out_shape, contiguous_size
 
 
-def device_reduce(ndarray x, op, out=None, bint keepdims=False):
+def device_reduce(ndarray x, int op, out=None, bint keepdims=False):
     cdef ndarray y, z
     cdef memory.MemoryPointer ws
-    cdef int dtype_id, ndim_out, kv_bytes
+    cdef int dtype_id, ndim_out, kv_bytes, x_size
```
Comment (jakirkham) [inline cupy/cuda/cub.pyx]: Shouldn't this also be `size_t`? Proposing changing the type in PR ( https://github.com/cupy/cupy/pull/3117 ).

**Pull request merged**